In [1]:
from sqlalchemy import create_engine, text

In [3]:
import pandas as pd

In [2]:
engine = create_engine('sqlite:///sql-murder-mystery.db')

In [4]:
query = '''
SELECT *
FROM sqlite_master
'''
pd.read_sql(query,con=engine)

type                                   name                tbl_name  \
0     table                     crime_scene_report      crime_scene_report   
1     table                        drivers_license         drivers_license   
2     table                                 person                  person   
3     table                 facebook_event_checkin  facebook_event_checkin   
4     table                              interview               interview   
5     table                     get_fit_now_member      get_fit_now_member   
6     index  sqlite_autoindex_get_fit_now_member_1      get_fit_now_member   
7     table                   get_fit_now_check_in    get_fit_now_check_in   
8     table                                 income                  income   
9     table                               solution                solution   
10  trigger                         check_solution                solution   

    rootpage                                                sql  
0          2  CREATE TABLE crime_scene_report (\n        dat...  
1          3  CREATE TABLE drivers_license (\n        id int...  
2          4  CREATE TABLE person (\n        id integer PRIM...  
3          5  CREATE TABLE facebook_event_checkin (\n       ...  
4          6  CREATE TABLE interview (\n        person_id in...  
5          7  CREATE TABLE get_fit_now_member (\n        id ...  
6          8                                               None  
7          9  CREATE TABLE get_fit_now_check_in (\n        m...  
8         10  CREATE TABLE income (\n        ssn integer PRI...  
9         11  CREATE TABLE solution (\n        user integer,...  
10         0  CREATE TRIGGER check_solution AFTER INSERT ON ...

In [5]:
query = '''
SELECT *
FROM crime_scene_report
'''
pd.read_sql(query,con=engine)

date     type                                        description  \
0     20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1     20180115   murder                  Life? Dont talk to me about life.   
2     20180115   murder  Mama, I killed a man, put a gun against his he...   
3     20180215   murder                         REDACTED REDACTED REDACTED   
4     20180215   murder  Someone killed the guard! He took an arrow to ...   
...        ...      ...                                                ...   
1223  20180430  bribery                                                 \n   
1224  20180430    fraud                  ‘Why not?’ said the March Hare.\n   
1225  20180430  assault                                                 \n   
1226  20180501  assault  be NO mistake about it: it was neither more no...   
1227  20180115   murder  Security footage shows that there were 2 witne...   

              city  
0              NYC  
1           Albany  
2             Reno  
3         SQL City  
4         SQL City  
...            ...  
1223  Garden Grove  
1224         Houma  
1225       Fontana  
1226       Trenton  
1227      SQL City  

[1228 rows x 4 columns]

Experienced SQL sleuths start here
A crime has taken place and the detective needs your help. The detective gave you the crime scene report, but you somehow lost it. You vaguely remember that the crime was a ​murder​ that occurred sometime on ​Jan.15, 2018​ and that it took place in ​SQL City​. Start by retrieving the corresponding crime scene report from the police department’s database.

In [20]:
query = '''
SELECT *
FROM crime_scene_report
WHERE city = "SQL City" and type = "murder" and date = 20180115
'''
pd.read_sql(query,con=engine).values

array([[20180115, 'murder',
        'Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".',
        'SQL City']], dtype=object)

array([[20180115, 'murder',
        'Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".',
        'SQL City']], dtype=object)

In [26]:
query = '''
SELECT *
FROM person
WHERE name LIKE "%Annabel%" and address_street_name LIKE "%Franklin Ave%"
'''
pd.read_sql(query,con=engine)

id            name  license_id  address_number address_street_name  \
0  16371  Annabel Miller      490173             103        Franklin Ave   

         ssn  
0  318771143

In [33]:
query = '''
SELECT *
FROM person
WHERE address_street_name LIKE "%Northwestern Dr%"
ORDER BY address_number DESC
'''
(pd.read_sql(query,con=engine))

id                  name  license_id  address_number  \
0   14887        Morty Schapiro      118009            4919   
1   17729        Lasonya Wildey      439686            3824   
2   53890        Sophie Tiberio      957671            3755   
3   73368        Torie Thalmann      773862            3697   
4   96595         Coretta Cubie      303645            3631   
5   19420           Cody Schiel      890431            3524   
6   93509        Emmitt Aceuedo      916706            3491   
7   87456   Leonora Wolfsberger      215868            3483   
8   36378        Freddie Ellzey      267882            3449   
9   53076           Boris Bijou      664914            3327   
10  28360        Rashad Cascone      924517            3212   
11  23044          Val Portlock      924989            3143   
12  51114     Christena Saffell      814284            3055   
13  68690            Yer Modest      852404            3046   
14  44004           Alison Eska      327797            2951   
15  98593           Jonah Toner      578982            2947   
16  59307          Cyril Yongue      753524            2903   
17  34155      Allyson Lazenson      814209            2840   
18  90397     Dionna Kranwinkle      797260            2678   
19  54564         Sammie Sengun      217286            2590   
20  53699      Christian Pagano      628433            2585   
21  34352    Vincenza Burkhardt      436470            2512   
22  80921       Donovan Yackley      222395            2386   
23  96178           Galina Khuu      450248            2183   
24  81885         Kurtis Masuda      934201            2108   
25  88423           Teri Ehrich      510876            1879   
26  51552           Bonita Knie      764124            1726   
27  72076     Courtney Bordeaux      324614            1707   
28  81052     Calvin Czelusniak      471483            1516   
29  71924        Laurie Scronce      774116            1418   
30  23960  Kristopher Lagerberg      658777            1392   
31  70485     Andera Kostelecky      354822            1252   
32  22239         Dusty Sigafus      710517            1125   
33  18376              Josh Shi      653712            1091   
34  29608       Willetta Lurvey      800568            1028   
35  33415         Homer Simpson      108392             981   
36  55130        Shiloh Huntoon      472118             951   
37  44542        Ivory Mcgrevey      873842             897   
38  99826          Ivy Kazarian      994470             892   
39  25615         Olevia Morena      388477             812   
40  40760          Renata Faver      976650             745   
41  10010             Muoi Cary      385336             741   
42  39688            Abe Roeker      537445             689   
43  26758      Angelena Billman      942074             629   
44  75484          Del Tacderen      447887             607   
45  12711        Norman Apolito      667757             599   
46  85280             Donn Duin      569122             540   
47  40336       Fredia Billeter      251082             404   
48  15171          Weldon Penso      336999             311   
49  89906       Kinsey Erickson      510019             309   

   address_street_name        ssn  
0      Northwestern Dr  111564949  
1      Northwestern Dr  917817122  
2      Northwestern Dr  442830147  
3      Northwestern Dr  341559436  
4      Northwestern Dr  378403829  
5      Northwestern Dr  947110049  
6      Northwestern Dr  979073160  
7      Northwestern Dr  565203106  
8      Northwestern Dr  474117596  
9      Northwestern Dr  401191868  
10     Northwestern Dr  838907424  
11     Northwestern Dr  100593316  
12     Northwestern Dr  247657547  
13     Northwestern Dr  419238039  
14     Northwestern Dr  306417867  
15     Northwestern Dr  745853548  
16     Northwestern Dr  866057839  
17     Northwestern Dr  991150469  
18     Northwestern Dr  278964807  
19     Northwestern Dr  848854310  
20     Northwestern Dr  644772304  
2

## Testigos : 

id	name	license_id	address_number	address_street_name	ssn
0	16371	Annabel Miller	490173	103	Franklin Ave	318771143


id	name	license_id	address_number	address_street_name	ssn
0	14887	Morty Schapiro	118009	4919	Northwestern Dr	111564949

In [40]:
query = '''
SELECT *
FROM interview
WHERE person_id = 16371 OR person_id = 14887

'''
pd.read_sql(query,con=engine).values

array([[14887,
        'I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".'],
       [16371,
        'I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.']],
      dtype=object)

## INTERVIEWS

        'I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".'],
       [16371,
        'I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.']],
      dtype=object)

In [45]:
# LOS 3 SOSPECHOSOS DE ANNABEL - PISTA = GYM
query = '''
SELECT *
FROM get_fit_now_member	
WHERE membership_status = "gold"
'''
pd.read_sql(query,con=engine)

id  person_id               name  membership_start_date  \
0   NL318      65076    Everette Koepke               20170926   
1   0YJ24      80651     Waneta Wellard               20171206   
2   5Y28Y      15218  Millicent Yessios               20180205   
3   A5N3S      24541       Mary Cameron               20170501   
4   R4J4S      13703         Lien Yasin               20170414   
..    ...        ...                ...                    ...   
63  0T622      98782    Sarina Overpeck               20171109   
64  BL3KC      24033      Collene Gulde               20170913   
65  48Z7A      28819       Joe Germuska               20160305   
66  48Z55      67318      Jeremy Bowers               20160101   
67  90081      16371     Annabel Miller               20160208   

   membership_status  
0               gold  
1               gold  
2               gold  
3               gold  
4               gold  
..               ...  
63              gold  
64              gold  
65              gold  
66              gold  
67              gold  

[68 rows x 5 columns]

In [59]:
# LOS 3 SOSPECHOSOS DE ANNABEL - PISTA = GYM
query = '''
SELECT *
FROM (get_fit_now_check_in)
WHERE check_in_date LIKE "%0901"
ORDER BY check_in_date DESC
'''
pd.read_sql(query,con=engine)

membership_id  check_in_date  check_in_time  check_out_time
0         VHN8M       20170901            513            1013
1         C581L       20170901            913             954

In [61]:
# LOS 3 SOSPECHOSOS DE ANNABEL - PISTA = GYM
query = '''
SELECT *
FROM get_fit_now_member	
WHERE id = "VHN8M" or id = "C581L" 
'''
pd.read_sql(query,con=engine)

id  person_id             name  membership_start_date membership_status
0  C581L      53262    Monty Torrico               20170307            silver
1  VHN8M      26473  Travis Bressler               20170516              gold

In [44]:
#LOS SOSPECHOSOS DE ANNABEL - PISTA = MATRICULA

query = '''
SELECT *
FROM drivers_license
WHERE plate_number LIKE "%H42W%"

'''
pd.read_sql(query,con=engine)

id  age  height eye_color hair_color  gender plate_number   car_make  \
0  183779   21      65      blue     blonde  female       H42W0X     Toyota   
1  423327   30      70     brown      brown    male       0H42W2  Chevrolet   
2  664760   21      71     black      black    male       4H42WR     Nissan   

  car_model  
0     Prius  
1  Spark LS  
2    Altima